File: src/explainable_engine/models/edge.py
from enum import Enum

from pydantic import BaseModel, Field


class TransformationType(str, Enum):
    WEIGHTED_SUM = "weighted_sum"
    NORMALIZATION = "normalization"
    THRESHOLD = "threshold"
    CUSTOM = "custom"


class Edge(BaseModel):
    source: str
    target: str
    weight: float = Field(ge=0.0, le=1.0)
    transformation_type: TransformationType = TransformationType.WEIGHTED_SUM
    metadata: dict | None = None

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "source": "trend_strength",
                    "target": "market_regime_score",
                    "weight": 0.4,
                    "transformation_type": "weighted_sum",
                }
            ]
        }
    }

File: internal/models/edge.go
package models

// TransformationType represents how values are transformed along an edge.
type TransformationType string

const (
	TransformationWeightedSum   TransformationType = "weighted_sum"
	TransformationNormalization TransformationType = "normalization"
	TransformationThreshold     TransformationType = "threshold"
	TransformationCustom        TransformationType = "custom"
)

// Edge represents a directed connection between two nodes.
type Edge struct {
	Source             string             `json:"source"`
	Target             string             `json:"target"`
	Weight             float64            `json:"weight"`
	TransformationType TransformationType `json:"transformation_type"`
	Metadata           map[string]string  `json:"metadata,omitempty"`
}
